README.md

```diff
@@ -1,2 +1,175 @@
+`/libp2p-circuit-relay/ipfs/QmFoo/libp2p-circuit-relay/ipfs/QmBar`
+
+Relay addresses can encapsulated in an /ipfs address to signal a specific relay node to use:
+`/ipfs/QmRelay/libp2p-circuit-relay/ipfs/QmFoo`
+or `/dns6/relay.libp2p.io/ipfs/QmRelay/libp2p-circuit-relay/ipfs/QmFoo`
+
+
+### /ipfs/relay/circuit stream protocol
+
+Each relayed connection corresponds to two multistreams,
+one between QmOne and QmRelay, the other between QmRelay and QmTwo.
+
+QmOne writes:
+
+```
+/ipfs/relay/circuit/0.1.0
+QmOne
+QmTwo
+inner connection data from here on, most likely multistream-select, then secio
+```
+
+```
+<multistream-header>
+<src>
+<dst>
+<data>
+```
+
+- `multistream-header`: This field is not strictly part of the relay stream.
+  - Multistreams are prefixed with a multicodec field identifying the protocol used for the stream.
+- `src`: The PeerID of the node opening the connection.
+- `dst`: The PeerID of the node that `src` wants to connect to.
+- `data`: Actual stream data, usually a multistream (or multistream-select).
+
+What happens:
+- QmRelay reads the header (not the data) from the stream.
+- Opens the stream to QmTwo, writes the header and shortcircuits the two streams.
+- QmOne and QmTwo do all the usual connection setup through the shortcircuited streams.
+
+Also:
+- QmRelay allows only one relayed connection between any two nodes.
+- QmRelay validates the `src` header field.
+
+Implementation:
+- QmTwo has listeners for the relay transport.
+  - Encapsulated: `/ipfs/QmRelay/libp2p-circuit-relay`
+    - Use that node for connecting through.
+    - Accept connections through it.
+  - `/libp2p-circuit-relay
+    - Accept connections through any relay node.
+  - These two can be conbined.
+- QmRelay has the `Swarm.EnableRelaying` config option enabled.
+
+### Interfaces
+
+```go
+import (
+  tpt "github.com/libp2p/go-libp2p-transport"
+  p2pnet "github.com/libp2p/go-libp2p-net"
+  peer "github.com/libp2p/go-libp2p-peer"
+  proto "github.com/libp2p/go-libp2p-protocol"
+)
+
+const ID proto.ID = "/ipfs/relay/circuit/0.1.0"
+
+type CircuitRelay interface {
+  tpt.Transport
+  p2pnet.StreamHandler
+
+  AllowRelaying(bool whetherToRelayOthersConnections)
+}
+```
+
+
+### Removing existing relay protocol
+
+Note that there is an existing swarm protocol colloqiually called relay.
+It lives in the go-libp2p package and is named `/ipfs/relay/line/0.1.0`.
+
+- Introduced in ipfs/go-ipfs#478 (28-Dec-2014).
+- No changes except for ipfs/go-ipfs@de50b2156299829c000b8d2df493b4c46e3f24e9.
+  - Changed to use multistream muxer.
+- Shortcomings
+  - No end-to-end encryption.
+  - No rate limiting (DoS by resource exhaustion).
+  - Doesn't verify src id in ReadHeader(), easy to fix.
+- Capable of *accepting* connections, and *relaying* connections.
+- *Not capable of *connecting* via relaying.*
+
+Since the existing protocol is incomplete, insecure, and certainly not used,
+we can safely remove it.
```

Comment:
Hm, we do need to be able to dial multiaddrs that the relaying node does not previously know. I think this is two things.

- (1) a circuit relay that dials other multiaddrs.
- (2) a circuit relay that understands `/{ipfs, p2p}/<peerid>` multiaddrs, and can dial to those with its own routing.

it would be useful to setup a bunch of boxes that do (1) kind of relay without dht or anything else. just dumb pipes.